src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+                throw new ArgumentOutOfRangeException($"{nameof(options)}.{nameof(GraphQLHttpMiddlewareOptions.WebSockets)}.{nameof(GraphQLWebSocketOptions.KeepAliveTimeout)}");
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+        }
+        Connection = sendStream ?? throw new ArgumentNullException(nameof(sendStream));
+        _cancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(sendStream.RequestAborted);
+        CancellationToken = _cancellationTokenSource.Token;
+    }
+
+    /// <inheritdoc/>
+    public virtual Task InitializeConnectionAsync()
+    {
+        var connectInitWaitTimeout = _options.WebSockets.ConnectionInitWaitTimeout ?? DefaultConnectionTimeout;
+        if (connectInitWaitTimeout != Timeout.InfiniteTimeSpan)
+        {
+            _ = Task.Run(async () =>
+            {
+                await Task.Delay(connectInitWaitTimeout, CancellationToken); // CancellationToken is set when this class is disposed
+                if (_initialized == 0)
+                    await OnConnectionInitWaitTimeoutAsync();
+            });
+        }
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Executes once the initialization timeout has expired without being initialized.
+    /// This method may execute concurrently with other code in this class.
+    /// </summary>
+    protected virtual Task OnConnectionInitWaitTimeoutAsync()
+        => ErrorConnectionInitializationTimeoutAsync();
+
+    /// <summary>
+    /// Called when the WebSocket connection (not necessarily the HTTP connection) has been terminated.
+    /// Disposes of all active subscriptions, cancels all existing requests,
+    /// and prevents any further responses.
+    /// </summary>
+    public virtual void Dispose()
+    {
+        var cts = Interlocked.Exchange(ref _cancellationTokenSource, null);
+        if (cts != null)
+        {
+            cts.Cancel();
+            cts.Dispose();
+            Subscriptions.Dispose();
+        }
+        GC.SuppressFinalize(this);
+    }
+
+    /// <summary>
+    /// Indicates if the connection has been initialized yet.
+    /// </summary>
+    protected bool Initialized
+        => _initialized == 1;
+
+    /// <summary>
+    /// Sets the initialized flag if it has not already been set.
+    /// Returns <see langword="false"/> if it was already set.
+    /// </summary>
+    protected bool TryInitialize()
+        => Interlocked.Exchange(ref _initialized, 1) == 0;
+
+    /// <summary>
+    /// Executes when a message has been received from the client.
+    /// </summary>
+    /// <exception cref="OperationCanceledException"/>
+    public abstract Task OnMessageReceivedAsync(OperationMessage message);
+
+    /// <summary>
+    /// Executes upon a request to close the connection from the client.
+    /// </summary>
+    protected virtual Task OnCloseConnectionAsync()
```

Comment:
I see only old protocol uses it reacting on `GQL_CONNECTION_TERMINATE`. What is alternative to terminate connection for the new protocol?

Reply:
The new protocol simply uses the websocket close message to gracefully terminate the connection.

Reply:
https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md

> The client closes the socket and the connection by dispatching a 1000: Normal Closure close event to the server indicating a normal closure.